sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosClientCache.scala

```diff
@@ -2,45 +2,65 @@
 // Licensed under the MIT License.
 package com.azure.cosmos.spark
 
-import com.azure.cosmos.implementation.{CosmosClientMetadataCachesSnapshot, SparkBridgeImplementationInternal}
-import com.azure.cosmos.{ConsistencyLevel, CosmosAsyncClient, CosmosClientBuilder, ThrottlingRetryOptions}
+import com.azure.cosmos.implementation.{CosmosClientMetadataCachesSnapshot, CosmosDaemonThreadFactory, SparkBridgeImplementationInternal}
+import com.azure.cosmos.spark.diagnostics.BasicLoggingTrait
+import com.azure.cosmos.{ConsistencyLevel, CosmosAsyncClient, CosmosClientBuilder, CosmosItemOperation, ThrottlingRetryOptions}
 import org.apache.spark.broadcast.Broadcast
 
-import java.time.Duration
+import java.time.{Duration, Instant}
 import java.util.ConcurrentModificationException
+import java.util.concurrent.{Executors, ScheduledExecutorService, TimeUnit}
+import java.util.concurrent.atomic.AtomicLong
 import scala.collection.concurrent.TrieMap
 
 // scalastyle:off underscore.import
 import scala.collection.JavaConverters._
 // scalastyle:on underscore.import
 
-private[spark] object CosmosClientCache {
-  private[this] val cache = new TrieMap[CosmosClientConfiguration, CosmosAsyncClient]
+private[spark] object CosmosClientCache extends BasicLoggingTrait {
+  private[this] val unusedClientTtlInMs = 5 * 60 * 1000
+  private[this] val cleanupIntervalInSeconds = 60
+  private[this] val cache = new TrieMap[CosmosClientConfiguration, CosmosClientCacheMetadata]
+  private[this] val executorService:ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor(
+    new CosmosDaemonThreadFactory("CosmosClientCache"));
+
+  this.executorService.scheduleWithFixedDelay(
+    () => this.onCleanup,
+    this.cleanupIntervalInSeconds,
+    this.cleanupIntervalInSeconds,
+    TimeUnit.SECONDS);
 
   def apply(cosmosClientConfiguration: CosmosClientConfiguration,
             cosmosClientStateHandle: Option[Broadcast[CosmosClientMetadataCachesSnapshot]]): CosmosAsyncClient = {
     cache.get(cosmosClientConfiguration) match {
-      case Some(client) => client
+      case Some(clientCacheMetadata) => {
+        clientCacheMetadata.lastRetrieved.set(Instant.now.toEpochMilli)
```

Comment:
so is the lastRetrieved will be set each time BulkWriter(or others) being initialized? since it is not marked by each request/response, how we make sure the client is not being actively used?